infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```
@@ -0,0 +1,316 @@
+            StateTransferBatchProcessor batchProcessor, Stream<TransferBatchRequest> batchStream) {
+        Result<Long, TransferSegmentFailure> accumulatedResult = Result.ok(0L);
+
+        Result<Long, TransferSegmentFailure> resultOfTransfer =
+                batchStream.reduce(accumulatedResult, (resultSoFar, nextBatch) -> {
+                    TransferBatchResponse transferBatchResponse =
+                            batchProcessor.transfer(nextBatch).join();
+                    if (transferBatchResponse.getStatus() == FailureStatus.FAILED) {
+                        return Result.error(new TransferSegmentFailure());
+                    } else {
+                        return resultSoFar
+                                .map(sumSoFar -> sumSoFar + transferBatchResponse
+                                        .getTransferBatchRequest()
+                                        .getAddresses()
+                                        .size());
+                    }
+                }, (oldSum, newSum) -> newSum);
+
+        return CompletableFuture.completedFuture(resultOfTransfer);
+    }
+
+
+    /**
+     * Based on a result of a state transfer, update a segment status.
+     *
+     * @param result      A result, containing a total number of records transferred or a failure.
+     * @param totalNeeded A total number of records that must be transferred.
+     * @return An updated segment status.
+     */
+    TransferSegmentStatus createStatusBasedOnTransferResult(
+            Result<Long, TransferSegmentFailure> result, long totalNeeded) {
+
+        Result<Long, TransferSegmentFailure> checkedResult = result.flatMap(totalTransferred -> {
+            if (totalTransferred != totalNeeded) {
+                String error = "Needed: " + totalNeeded + ", but transferred: " + totalTransferred;
+
+                return Result.error(new TransferSegmentFailure(new IllegalStateException(error)));
```

Is this overwriting the original error? You got an error during transfer, but when you come into this method, the first thing you do is to check total transferred, but if a network/wrong epoch exception happens and transfer failed, you always swallow that error and transform it into a IllegalStateException?

---

Fixed.